src/emailcontactmodel.cpp

```diff
@@ -0,0 +1,148 @@
+ * Apache License, version 2.0.  The full text of the Apache License is at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+#include "emailcontactmodel.h"
+#include <QHash>
+
+
+namespace  {
+    const int MAX_COUNT = 100;
+}
+
+
+EmailContactModel::EmailContactModel(QAbstractItemModel *parent)
+    : QAbstractListModel(parent)
+{
+    initialize(MAX_COUNT);
+}
+
+
+static void push(QHash<QMailAddress, QMailTimeStamp> &container, const QMailAddress &address, const QMailTimeStamp &stamp)
+{
+    if (!address.isEmailAddress())
+        return;
+
+    QHash<QMailAddress, QMailTimeStamp>::iterator i = container.find(address);
+    if (i == container.end()) {
+        container.insert(address, stamp);
+    } else {
+        if (i.value().toLocalTime() < stamp.toLocalTime())
+            i.value() = stamp;
+    }
+}
+
+
+static bool comparer(const QPair<QMailTimeStamp, QMailAddress> &value1, const QPair<QMailTimeStamp, QMailAddress> &value2)
+{
+    return value1.first < value2.first;
+}
+
+
+void EmailContactModel::initialize(const int maxCount)
+{
+    QHash<QMailAddress, QMailTimeStamp> container;
+
+    QMailStore *mailStore = QMailStore::instance();
+    Q_ASSERT(mailStore);
+
+    if (QMailStore::initializationState() != QMailStore::Initialized)
+        return;
+
+    QMailMessageKey incomingKey(QMailMessageKey::status(QMailMessage::Incoming) & ~QMailMessageKey::status(QMailMessage::Trash));
+    QMailMessageIdList incomingIds = mailStore->queryMessages(incomingKey);
+    int count = 0;
+    for (const auto& id: incomingIds) {        
+        if (count++ == maxCount)
+            break;
+
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QMailAddress address = metaData.from();
+        push(container, address, timeStamp);        
+    }
+
+    QMailMessageKey outgoingKey(QMailMessageKey::status(QMailMessage::Outgoing));
+    QMailMessageIdList outgoingIds = mailStore->queryMessages(outgoingKey);
+    count = 0;
+    for (const auto& id: outgoingIds) {
+        if (count++ == maxCount)
+            break;
+
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QList<QMailAddress> recipients = metaData.recipients();
+        for (const auto& address : recipients)
+            push(container, address, timeStamp);
+    }
+
+    m_container.clear();
+    for (QHash<QMailAddress, QMailTimeStamp>::const_iterator i = container.begin(); i != container.end(); ++i) {
+        m_container.push_back(qMakePair(i.value(), i.key()));
+    }
+    qSort(m_container.begin(), m_container.end(), comparer);
```

Comment:
qSort is obsoleted in favor of std::sort. Could already use that instead.

Reply:
good idea
std::somethinks I like more
done